src/hooks/useForm.js

```diff
@@ -0,0 +1,44 @@
+import { useState, useCallback } from 'react';
+
+const useForm = (
+  { initialValues, onSubmit, validator = () => {}, validateOnChange = false },
+  ...dependencies
+) => {
+  const [values, setValues] = useState(initialValues);
+  const [errors, setErrors] = useState({});
+
+  const handleSubmit = useCallback(() => {
+    const newErrors = validator(values) || {};
+    const valid = !Object.values(newErrors).flat().length;
+    if (valid) {
+      onSubmit(values);
+    } else {
+      setErrors(newErrors);
+    }
+  }, [values, errors, ...dependencies]);
+
+  const runValidations = useCallback(
+    newValues => {
+      if (validateOnChange) {
+        setErrors(validator(newValues));
+      }
+    },
+    [validateOnChange, validator],
+  );
+
+  const handleValueChange = useCallback(
+    (key, value) => {
+      const newValues = {
+        ...values,
+        [key]: value,
+      };
+      setValues(newValues);
+      runValidations(newValues);
+    },
+    [values, runValidations],
+  );
+
+  return [values, setValues, errors, setErrors, handleValueChange, handleSubmit];
```

Comment:
What about an object? I think this array is a bit overwhelming.

Reply:
Yep, sure, I was convinced the standard was to return arrays, but seems that's not the case, so no problem